keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+}
+
+// ResourceClaim is created by users to describe which resources they need.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+type ResourceClaim struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
+	// the resource. It removes the finalizer again when a) the allocation
+	// attempt has definitely failed or b) when the allocated resource was
+	// deallocated. This ensures that resources are not leaked.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// ResourceClaim.
+	Spec core.ResourceClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ResourceClaimStatus
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// ResourceClassName references the driver and additional parameters
+	// via the name of a ResourceClass that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists, but a
+	// driver-specific admission webhook may require that and is allowed to
+	// reject claims where the class is missing.
+	ResourceClassName string
+
+	// Parameters references a separate object with arbitrary parameters
+	// that will be used by the
+	// driver when allocating a resource for the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	Parameters ResourceClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod (AllocationModeDelayed). Other modes
+// might get added in the future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeDelayed, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeDelayed AllocationMode = "Delayed"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
```

Comment:
I can't tell if this hasn't been updated to reflect `PodScheduling` or if I am missing something.  It feels like most of this stanza could be replaced with a `Conditions`

`driverName` is informational, so I could maybe see keeping that, but without the params it is not very useful or actionable.

`scheduling` is supposed to be moved out, right?

`allocation` is interesting only for the nodeSelector (optional hint).  I'm not sure whether the `sharedResource` flag is useful yet, given how problematic `reservedFor` is.

`deallocate` is supposed to be moved out, right?  I really dislike an imperative thing like that and I will try to avoid it.

I'll talk about `reservedFor` in the thread below


From a high level, let's think about what we need:

* A way for a driver to indicate "allocation is complete"
* A way for a driver to hint "these nodes support the allocation"
* A way to indicate errors, e.g. "something about this allocation can not be satisfied"
* MAYBE a way for end-users to retrieve information about the resource, e.g. the params used to create)

Anything else?  You seem very concerned with the multi-scheduler case, but I don't think I have ever seen that in the wild.

Reply:
> driverName is informational, so I could maybe see keeping that, but without the params it is not very useful or actionable.

kubelet needs it when the resource is allocated and the ResourceClass is gone. Without a copy of the driver name it won't know which plugin to call.

> scheduling is supposed to be moved out, right?

I am trying.

> allocation is interesting only for the nodeSelector (optional hint).

It is also relevant for the scheduler. It needs to know whether it still needs to trigger allocation ("delayed allocation" case) or wait for it to finish before reserving the claim ("immediate allocation" case).

> deallocate is supposed to be moved out, right?

As with "scheduling", I am trying.

I don't want to go into details, but the problem with handling scheduling separately from tracking the state of a claim is that it is no longer obvious whether a claim is in the process of being allocated or deallocated, which affects reserving it.

> You seem very concerned with the multi-scheduler case, but I don't think I have ever seen that in the wild.

I've heard of people doing it. I used the multi-scheduler case to make it clear that the scheduler should schedule pods independently from each other.

If we assume a single scheduler, then it could detect that it has two pods which share a claim and then decide for which it starts scheduling. Do we allow it to start with both? Then we still have potential conflicts.

Reply:
> I am trying.

See https://github.com/pohly/enhancements/pull/13/files for the result and the opens around it.

Reply:
>> driverName is informational, so I could maybe see keeping that, but without the params it is not very useful or actionable.
> kubelet needs it when the resource is allocated and the ResourceClass is gone. Without a copy of the driver name it won't know which plugin to call.

OK, fair.  Does the on-node driver have enough information here to do its job?

Reply:
> See https://github.com/pohly/enhancements/pull/13/files for the result and the opens around it.

Will do

Reply:
> Does the on-node driver have enough information here to do its job?

It depends on how much it puts into `Allocation.Attributes` (aka Allocation.VolumeHandle` after renaming it). If it references some additional information there, then it will also need that, otherwise it has everything immediately.